doc/experimental_ui_api.md

```diff
@@ -0,0 +1,2027 @@
+a comprehensive description of how style properties affect the layout and
+display of boxes is also discussed in the [Layout and visuals] section.
+
+### States
+
+The list of states is as follows, from highest precedence to lowest:
+
+* `disabled`: The box is disabled, which means that user interaction with it
+  does nothing.
+* `pressed`: The left mouse button was pressed down inside the boundaries of
+  the box, and has not yet been released.
+* `hovered`: The mouse cursor is currently inside the boundaries of the box.
+* `selected`: The box is currently selected in some way, e.g. a checkbox is
+  checked or a list item is selected.
+* `focused`: The box currently has keyboard focus.
+
+The `pressed` and `hovered` states are properties of the box itself, and hence
+only apply to dynamic boxes. Any dynamic box can be hovered and pressed,
+although different boxes may have different requirements for what constitutes
+being pressed.
+
+The `disabled`, `selected`, and `focused` states are often shared among
+multiple boxes, and hence may apply to both static and dynamic boxes. For
+example, when an individual scrollbar is disabled, every box within that
+scrollbar has the `disabled` state.
+
+By default, it should be assumed that the `focused` state applies to every box
+within the element if the element has focus. Moreover, if the element has a
+`disabled` property, then the `disabled` state will also apply to every box
+when that property is set. Boxes that have different behavior document how they
+behave instead.
+
+### State cascading
+
+States fully cascade over each other. For instance, if there are styles that
+give hovered buttons yellow text and pressed buttons blue backgrounds, then a
+hovered and pressed button will have yellow text on a blue background.
+
+```lua
+-- Creating two styles with separate states like so...
+ui.Style "button$hovered" {
+    text_color = "yellow",
+}
+ui.Style "button$pressed" {
+    box_fill = "blue",
+}
+
+-- ...automatically implies the following cascaded style as well:
+ui.Style "button$hovered$pressed" {
+    text_color = "yellow",
+    box_fill = "blue",
+}
+```
+
+However, if an element is currently in multiple states, then states with higher
+precedence will override states with lower precedences. For instance, if one
+style makes hovered buttons red and another makes pressed buttons blue, then a
+button that is simultaneously hovered and pressed will be blue.
+
+```lua
+-- One style makes hovered buttons red, and another makes pressed buttons blue.
+ui.Style "button$hovered" {
+    box_fill = "red",
+}
+ui.Style "button$pressed" {
+    box_fill = "blue",
+}
+
+-- Then, the implicit cascaded style will make buttons blue:
+ui.Style "button$hovered$pressed" {
+    box_fill = "blue",
+}
+```
+
+Lastly, state precedences combine. A style for pressed and hovered buttons will
+override styles for only pressed or only hovered buttons. The highest state
```

Comment:
So basically lexicographical ordering based on state precedences?

Reply:
Basically, yes, where higher precedence means listed later in lexicographical ordering.